src/web/tools/EventManager.ts

```diff
@@ -0,0 +1,176 @@
+        event,
+        EventTypes.MOVE
+      );
+
+      const inBounds: boolean = this.isPointerInBounds({
+        x: adaptedEvent.x,
+        y: adaptedEvent.y,
+      });
+
+      const pointerIndex: number = this.activePointers.indexOf(
+        adaptedEvent.pointerId
+      );
+
+      if (inBounds) {
+        if (pointerIndex < 0) {
+          adaptedEvent.eventType = EventTypes.ENTER;
+          this.onPointerEnter(adaptedEvent);
+          this.addActivePointer(adaptedEvent.pointerId);
+        } else {
+          this.onPointerMove(adaptedEvent);
+        }
+      } else {
+        if (pointerIndex >= 0) {
+          adaptedEvent.eventType = EventTypes.OUT;
+          this.onPointerOut(adaptedEvent);
+          this.removeActivePointer(adaptedEvent.pointerId);
+        } else {
+          this.onPointerOutOfBounds(adaptedEvent);
+        }
+      }
+    });
+
+    this.view.addEventListener('pointercancel', (event: PointerEvent): void => {
+      event.preventDefault();
+
+      const adaptedEvent: AdaptedPointerEvent = this.mapEvent(
+        event,
+        EventTypes.CANCEL
+      );
+
+      this.onPointerCancel(adaptedEvent);
+    });
+  }
+
+  private onPointerDown(_event: AdaptedPointerEvent): void {}
+  private onPointerUp(_event: AdaptedPointerEvent): void {}
+  private onPointerMove(_event: AdaptedPointerEvent): void {}
+  private onPointerOut(_event: AdaptedPointerEvent): void {}
+  private onPointerEnter(_event: AdaptedPointerEvent): void {}
+  private onPointerCancel(_event: AdaptedPointerEvent): void {}
+  private onPointerOutOfBounds(_event: AdaptedPointerEvent): void {}
+
+  public setOnDownAction(callback: (event: AdaptedPointerEvent) => void): void {
+    this.onPointerDown = callback;
+  }
+  public setOnUpAction(callback: (event: AdaptedPointerEvent) => void): void {
+    this.onPointerUp = callback;
+  }
+  public setOnMoveAction(callback: (event: AdaptedPointerEvent) => void): void {
+    this.onPointerMove = callback;
+  }
+  public setOnOutAction(callback: (event: AdaptedPointerEvent) => void): void {
+    this.onPointerOut = callback;
+  }
+  public setOnEnterAction(
+    callback: (event: AdaptedPointerEvent) => void
+  ): void {
+    this.onPointerEnter = callback;
+  }
+  public setOnCancelAction(
+    callback: (event: AdaptedPointerEvent) => void
+  ): void {
+    this.onPointerCancel = callback;
+  }
+  public setOutOfBoundsAction(
+    callback: (event: AdaptedPointerEvent) => void
+  ): void {
+    this.onPointerOutOfBounds = callback;
+  }
+
+  private mapEvent(
+    event: PointerEvent,
+    eventType: EventTypes
+  ): AdaptedPointerEvent {
+    return {
+      x: event.clientX,
+      y: event.clientY,
+      offsetX: event.offsetX,
+      offsetY: event.offsetY,
+      pointerId: event.pointerId,
+      eventType: eventType,
+      pointerType: event.pointerType,
+      buttons: event.buttons,
+      time: event.timeStamp,
+    };
+  }
```

Comment:
Since you have access to `activePointers` in this method, wouldn't it make sense to distinguish between `UP`/`DOWN` and `POINTER_UP`/`POINTER_DOWN` here? At the moment those are only used in `ScaleGestureDetector` which sets those itself, but similar logic exists in other places as well.

Reply:
This one is a bit more complicated, since `activePointers` doesn't have the pointers all the time. It is used to mark whether the pointer is in bounds of view or not (and that's why it probably isn't the best name for that variable). I solved that with similar approach in another branch, which includes more changes in `EventManager`, such as adding TouchEvents